x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/ApiKeyGenerator.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authc.support;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyResponse;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.support.DLSRoleQueryValidator;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+
+import java.util.Arrays;
+import java.util.HashSet;
+
+public class ApiKeyGenerator {
+
+    private final ApiKeyService apiKeyService;
+    private final CompositeRolesStore rolesStore;
+    private final NamedXContentRegistry xContentRegistry;
+
+    public ApiKeyGenerator(ApiKeyService apiKeyService, CompositeRolesStore rolesStore, NamedXContentRegistry xContentRegistry) {
+        this.apiKeyService = apiKeyService;
+        this.rolesStore = rolesStore;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    public void generateApiKey(Authentication authentication, CreateApiKeyRequest request, ActionListener<CreateApiKeyResponse> listener) {
+        if (authentication == null) {
+            listener.onFailure(new ElasticsearchSecurityException("no authentication available to generate API key"));
+            return;
+        }
+        apiKeyService.ensureEnabled();
```

Comment:
This is called in `ApiKeyService#createApiKey()`. Is it intentional that we do it here too becayse `getRoleDescriptors` can be expensive ? If so can we also /instead inject the ApiKeyService in the Transport service and short circuit there ?

Reply:
Partly the relative expense, but partly the "right" error message. If the API key service is disabled you want to prioritise that error message ahead of any role format issues.
(That's specifically what triggered me to write it, I ran into something along those lines while testing)

Reply:
> If so can we also /instead inject the ApiKeyService in the Transport service and short circuit there

I don't know if we want to. I think we should prioritise authentication problems (even secondary auth problems) over the service being disabled. Happy to discuss if you feel otherwise.

Reply:
> , but partly the "right" error message.

Makes sense. 

> I think we should prioritise authentication problems 

Agreed, I didn't think this through